python/cudf/cudf/core/join/join.py

```diff
@@ -234,14 +234,57 @@ def _compute_join_keys(self):
                         for on in _coerce_to_tuple(self.right_on)
                     ]
                 )
+        elif self.on:
+            on_names = _coerce_to_tuple(self.on)
+            for on in on_names:
+                # If `on` if provided, checks whether merging on
+                # columns, indexes or merging  index with column
+                if (
+                    on in self.lhs._data.keys()
+                    and on not in self.rhs._data.keys()
+                ):
+                    # case1: merge on lhs column with rhs index
+                    if on in self.rhs.index.names:
+                        left_keys.extend([_Indexer(name=on, column=True)])
+                        right_keys.extend([_Indexer(name=on, index=True)])
+
+                elif (
+                    on not in self.lhs._data.keys()
+                    and on in self.rhs._data.keys()
+                ):
+                    # case2: merge on rhs column with lhs index
+                    if on in self.lhs.index.names:
+                        left_keys.extend([_Indexer(name=on, index=True)])
+                        right_keys.extend([_Indexer(name=on, column=True)])
+
+                elif (
+                    on not in self.lhs._data.keys()
+                    and on not in self.rhs._data.keys()
+                ):
+                    # case3: merge on lhs index with rhs index
+                    if (
+                        on in self.lhs.index.names
+                        and on in self.rhs.index.names
+                    ):
+                        left_keys.extend([_Indexer(name=on, index=True)])
+                        right_keys.extend([_Indexer(name=on, index=True)])
+                    else:
+                        # if none of the cases, then invalid key "on"
+                        # not found in both frames
+                        raise KeyError(on)
+
+                else:
+                    # case4: merge on lhs column with rhs column
+                    left_keys = [
+                        _Indexer(name=on, column=True) for on in on_names
+                    ]
+                    right_keys = [
+                        _Indexer(name=on, column=True) for on in on_names
+                    ]
```

Comment:
Even in here it's still possible `on` includes names not one of the column names. Maybe a better design is to check if `on` is valid before all if-elses?

Reply:
Thanks for pointing to EAFP. I think I would agree not to add overhead if `_Indexer` already does the check. One little nit I had with this is that the code block exist in a for-loop, and the `_Indexer` checks lazily not until `get` is called. 

The worst situation of this is: the first element in `on_names` is invalid, yet the algorithm still have to iterate over all `on_names`. Would some kind of early-termination help here?

Reply:
The `KeyError` will be raised on the first entry in `on` that doesn't exist in either an index column or a data column; isn't that the best case?

Reply:
Connected offline, for the performance concern addressed here, I opt for cleaner code logic (fewer if-elses) as opposed to early termination. As with wide table has never been a performance promise by cudf.

Reply:
Quick note: in general, the choice between EAFP and LBYL needs to be made on a case-by-case basis depending on 1) how frequent the exception cases are, and 2) how expensive the predicate is to evaluate. Python makes EAFP cheap if the exceptions don't occur frequently, but it does requiring thinking about how expensive the conditions are to check. Using EAFP as the _default_ can also be problematic if the "exceptional" cases are actually not that exceptional.